Enhance #31 left click outside of a menu or Esc to leave it.

JimZhouZZY opened the pull request:
Add feature: 
1. Left clicking outside a menu to leave a menu.
2. Press Escape key to leave a menu.

Add interface for key operation. Now we can add key binding in "level_scene.py" like this:
```python
...
 def key_down(self, keyname):
        ...
        if keyname == pygame.K_ESCAPE:
            self.leave_menu()
...
```

Grimmys commented on src/services/scene_manager.py:
```diff
@@ -49,6 +49,8 @@ def process_game_iteration(self) -> bool:
                 if event.button in (1, 3):
                     self.active_scene.button_down(event.button, event.pos)
+            elif event.type == pygame.KEYDOWN:
+                self.active_scene.key_down(event.key)
```
Looks good 👍

Grimmys commented on src/scenes/level_scene.py:
```diff
@@ -1971,6 +1971,49 @@ def new_turn(self) -> None:
 
+    def leave_menu(self):
+        # This part I copied the code from function self.right_click
+        # to prevent being infected by future changes of function self.right_click
```
I get your point, but what about code duplication?

Also, you for example fixed the cancelling of gold trades after cancelling the turn of a player, but since this code is not the same, the gold trade cancelling will happen only while cancelling with right click, not with left click/escape.
Code duplication should be avoid for this kind of situation: the different pieces of the project become harder to maintain on sync.

My suggestion: call this method inside of **right click** instead of having the same piece of code twice.
I will later try to refactor even more this piece of code because it's starting to get big, but that's another story.

Grimmys commented:
Feature is working well, good job!

Just waiting for the small refactor to be done.

JimZhouZZY commented:
I'm so sorry I made this ill-advised edit. I realised how stupid is was. And I modified the code seriously.

Here are the modified functions on clicks/keys.
+ `Left Click` outside will close a menu whose position is not fixed such as `Main menu`, `Item sell menu`...but except `Select an action` menu.
+ `Right Click` will function as before, no changes.
+ `Escape` will close all menus except `Select an action` menu.

And Replying to:
>My suggestion: call this method inside of right click instead of having the same piece of code twice.

I deleted the method `leave_menu`, because it doesn't have a good practicality.

Grimmys commented on src/scenes/level_scene.py:
```diff
@@ -1980,6 +1980,20 @@ def left_click(self, position: Position) -> None:
+                    self.menu_manager.active_menu._InfoBox__size[1]
+                )
+                if (
+                    self.menu_manager.active_menu is not None
```
You don't need this check.
Line 1981 is checking if the `active_menu` is defined, this whole code block is not going to be executed if `active_menu` is not defined.

JimZhouZZY replied:
It's my fault. Thank you for your patience and care to a beginner.

Grimmys commented on src/scenes/level_scene.py:
```diff
@@ -2198,6 +2212,19 @@ def button_down(self, button: int, position: Position) -> None:
+        """
+        if keyname == pygame.K_ESCAPE:
+            if (self.menu_manager.active_menu is not None
+                and self.menu_manager.active_menu.title != "Select an action"
```
Good idea to block such menus, but I don't think it's ever a good idea to rely on strings displayed on the screen.
Because it means that if one day someone wants to make a subtle change to this text, the application would be broken. Also, what about the multilingual support you are going to introduce? It may not be this text on the screen, but its translation.

I suggest two alternative solutions:
* `Infobox` has attribute `has_close_button`, we could assume that if a modal doesn't have any close button, then it means it cannot be close, because it's a blocking menu (waiting for user action). That would be the generic approach.
* `Infobox` also has attribute `identifier`, this identifier could be used to refer to the specific "select an action" modal instead of relying on something that may change. So the solution would be to give a meaningful id at the creation of the modal, and at this place check if the `active_menu` matches the given id. That would be a more specific approach.

Btw, if you need any help regarding how the UI components are working: [https://pygame-popup-manager.readthedocs.io/en/latest/api/components/info_box.html](https://pygame-popup-manager.readthedocs.io/en/latest/api/components/info_box.html)

JimZhouZZY replied:
Conditional judgement in Line2225 copied the original judgement in Line2132(in function right_click). I had no intention of resolving the above TODO at the time, so I copied it directly. Maybe we can handle it later. 

Another fact is that when multilingual support added, it will not be an issue cuz the code is going to be like this:
```python
if self.menu_manager.active_menu.title == STR_SELECT_AN_ACTION:
```

Grimmys replied:
My bad, I didn't notice it was already this way before!
This TODO has to be from a time where  `InfoBox` didn't have any id haha.

You are definitely right then, it should be handled in another PR since it is not introduced by this change.
I'm going to take care of it.

Grimmys commented:
> I'm so sorry I made this ill-advised edit. I realised how stupid is was. And I modified the code seriously.

No no no and definitely no, you don't have to be sorry for the work you are producing. In development, everyone makes "mistakes" all the time, even seniors. And what may be considered an error by someone might not be for someone else. Everything is about point of view! Solutions should be debated, and we need to think about whether or not what we did make sense, but  not blame ourselves and feel bad about it.

JimZhouZZY commented:
> > I'm so sorry I made this ill-advised edit. I realised how stupid is was. And I modified the code seriously.
> 
> No no no and definitely no, you don't have to be sorry for the work you are producing. In development, everyone makes "mistakes" all the time, even seniors. And what may be considered an error by someone might not be for someone else. Everything is about point of view! Solutions should be debated, and we need to think about whether or not what we did make sense, but not blame ourselves and feel bad about it.

Thank you for your kind words. I appreciate your encouragement and support. I’m still learning and trying to improve my skills, so it’s good to know that the community is so accommodating and easy-going. I’ll keep your advice in mind and try to do better next time. Thanks again!

JimZhouZZY commented:
I'm confused with the Unit tests. Sometimes it passes, but sometimes fails. Always on the line `self.assertIsInstance(self.start_screen.level, self.level_class)`. 
```
======================================================================
FAIL: test_load_existent_save (tests.test_start_screen.TestStartScreen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/rpg_tactical_fantasy_game/rpg_tactical_fantasy_game/tests/test_start_screen.py", line 147, in test_load_existent_save
    self.assertIsInstance(self.start_screen.level, self.level_class)
AssertionError: None is not an instance of <class 'src.scenes.level_scene.LevelScene'>

----------------------------------------------------------------------
Ran 110 tests in 27.345s
```

Grimmys commented:
Yes I noticed it, but it's passing most of the time, and the test doesn't look related to the changes you did.

I'm assuming it's just a flakky test (i.e., test wrongly failling sometimes) because the position is randomly generated.
Will investigate it later.

Grimmys reviewed (approved):
LGTM, thanks for the contribution
